[feat] Allow users define custom log handlers and attach them to the framework

dmargala opened the pull request:
This PR adds support for defining a custom log handler that can be specified in the reframe configuration file. 

For example, adding a plugin log handler would look something like the following:

```python
class PlugInLogHandler(logging.Handler):
    def emit(self, record):
        custom_logger.log(record)

def _create_plugin_handler(site_config, handlers_group):
    return PlugInLogHandler()

from reframe.core.logging import register_plugin_handler
register_plugin_handler(_create_plugin_handler)
```

There is some prior art in the reframe code base for a similar capability for custom launchers (see [extending-the-framework](https://reframe-hpc.readthedocs.io/en/stable/howto.html#extending-the-framework) section of the docs). There may be some other ways to implement this by extending the existing Stream or File log handlers to accept custom stream or file objects.

closes #3270

pep8speaks commented:
Hello @dmargala, Thank you for submitting the Pull Request!

* In the file [`reframe/core/logging.py`](https://github.com/reframe-hpc/reframe/blob/ca1ca762a86103de19dc7a4937601d7230a971cc/reframe/core/logging.py):

> [Line 690:1](https://github.com/reframe-hpc/reframe/blob/ca1ca762a86103de19dc7a4937601d7230a971cc/reframe/core/logging.py#L690): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

Do see the ReFrame [Coding Style Guide](https://github.com/reframe-hpc/reframe/wiki/coding-style-guide)

vkarak commented on reframe/core/logging.py:
```diff
@@ -678,39 +678,35 @@ def emit(self, record):
+}
+
+
+def register_plugin_handler(create_plugin_handler):
+    _create_handler_registry["plugin"] = create_plugin_handler
```
I would make this a decorator taking the name of the handler to register that will populate the above registry upon definition of each handler. This registry could start empty here and every log handler creation function (even the builtin ones) could be defined as

```python
@register_log_handler('handler_name')
def _create_foo_handler(site_config, handlers_group):
    ...
```

The nice thing with the decorator is that we can also do some validation of the function signature if we want to.

```suggestion
def register_log_handler(name):
    def _deco(fn):
        _create_handler_registry[name] = fn
        return fn
     
     return deco
```

vkarak reviewed (changes requested):
Thanks for the PR! I have some suggestions on how to make this idea more in sync with the similar idea of custom launchers and schedulers.

We should also add a small documentation section about it in the "how to".

dmargala commented on reframe/core/logging.py:
```diff
@@ -678,39 +678,35 @@ def emit(self, record):
-        else:
+        except KeyError:
             # Should not enter here
             raise AssertionError(f'unknown handler type: {handler_type}')
```
Done. I also changed `AssertionError` to `ConfigError` and dropped the "should not enter here" comment since we could enter there if a type is specified in the config that hasn't been registered.

vkarak commented on reframe/core/logging.py:
```diff
@@ -412,6 +412,19 @@ def stream_handler_kind(handler):
+    '''Register the decorated log handler creation function'''
+    def _create_handler_wrapper(fn):
+        _create_handlers[name] = fn
+        return fn
```
Just annoying code style nitpicking: leave a blank line after the return here.

vkarak reviewed (changes requested):
Lgtm, just some style nitpicking.

dmargala commented:
Thanks. I applied suggestions and also fixed indent spacing in the examples (the spacing in the code blocks had inherited the 3-space indents, I changed it to 4-spaces.

vkarak reviewed (approved):
Thanks!
